src/dynamic_programming/knapsack.md

```diff
@@ -0,0 +1,179 @@
+---
+tags:
+  - Original
+---
+
+# Knapsack problem
+Prerequisite knowledge: [Introduction to Dynamic Programming](https://cp-algorithms.com/dynamic_programming/intro-to-dp.html)
+
+## Introduction
+Consider the following example:
+
+### [[USACO07 Dec] Charm Bracelet](https://www.acmicpc.net/problem/6144) 
+There are $n$ distinct items and a knapsack of capacity $W$. Each item has 2 attributes, weight ($w_{i}$) and value ($v_{i}$). 
+You have to select a subset of items to put into the knapsack such that the total weight does not exceed the capacity $W$ and the total value is maximized.
+
+In the example above, since each object has only two possible states (taken or not taken),
+correspoding to binary 0 and 1. Thus, this type of problem is called "0-1 knapsack problem".
+
+## 0-1 Knapsack
+
+### Explaination
+
+In the example above, the input to the problem is the following: the weight of $i^{th}$ item $w_{i}$, value of $i^{th}$ item $v_{i}$, and the total capacity of the knapsack $W$.
+
+Let $f_{i, j}$ be the dynamic programming state holding the maximum total value the knapsack can carry with capacity $j$, when only the first $i$ items are considered.
+
+Assuming that all states of the first $i-1$ items have been processed, what are the options for the $i^{th}$ item?
+
+- When it is not put into the knapsack, the remaining capacity remains unchanged and total value does not change. Therefore, the maximum value in this case is $f_{i-1, j}$
+- When it is put into the knapsack, the remaining capacity decreases by $w_{i}$ and the total value increases by $v_{i}$,
+so the maximum value in this case is $f_{i-1, j-w_i} + v_i$
+
+From this we can derive the dp transition equation:
+
+$$f_{i, j} = \max(f_{i-1, j}, f_{i-1, j-w_i} + v_i)$$
+
+Further, as $f_{i}$ is only dependent on $f_{i-1}$, we can remove the first dimension. We obtain:
+
+$$f_j \gets \max(f_j, f_{j-w_i}+v_i)$$
+
+**It is important to understand this transition rule, because most of the transitions for knapsack problems are derived in a similar way.**
+
+### Implementation
+
+The alorithm described can be implemented in $O(nW)$ as:
+
+```.c++
+for (int i = 1; i <= n; i++)
+  for (int j = W; j >= w[i]; j--)
+    f[j] = max(f[j], f[j - w[i]] + v[i]);
+```
+
+Note the order of enumeration. We go over all possible weights for each item rather than the other way round. If we execute the loops in the other order, $f_W$ will get updated assuming that we can pick only one item.
```

Comment:
```suggestion
Again, note the order of execution. It should be strictly followed to ensure the following invariant: Right before the pair $(i, j)$ is processed, $f_k$ corresponds to $f_{i,k}$ for $k > j$, but to $f_{i-1,k}$ for $k < j$. This ensures that $f_{j-w_i}$ is taken from the $(i-1)$-th step, rather than from the $i$-th one.
```